### PR TITLE
tan + sin funcs

### LINKED ISSUE #384: Add trigonometric function operators for F24
**Is your feature request related to a problem? Please describe.**
A request by a Bend user in this issue: https://github.com/HigherOrderCO/Bend/issues/570.

We have log, atan2 and pow for floats, but not trigonometric functions like sin, cos and tan.
I think they're common enough to warrant being a native operation. We should have at least one of them, the others could be easily derived (cos(x) = sqrt(1 - sin(x)²), etc)

**Describe the solution you'd like**
Add at least one of `sin`, `cos` and `tan` functions as a native floating point operation. We could use `SHL` and `SHR` tags to implement them for f24 types.

**Describe alternatives you've considered**
Approximating them using the already present operations. I think it can be done in around ~10-20 interactions each.

### PR DESCRIPTION
**Pull request by enricozb**
fixes https://github.com/HigherOrderCO/HVM/issues/384

### CODE REVIEW

File: src/hvm.c [hunk 1]
```diff
@@ -570,6 +570,10 @@ static inline Numb operate(Numb a, Numb b) {
         case OP_AND: return new_f24(atan2f(av, bv));
         case OP_OR:  return new_f24(logf(bv) / logf(av));
         case OP_XOR: return new_f24(powf(av, bv));
+        case OP_SHL: return new_f24(sin(av + bv));
+        case FP_SHL: return new_f24(sin(av + bv));
```
Comment (developedby) [inline src/hvm.c:574]: We don't need to waste f24 FP_SHL and FP_SHR here

File: src/hvm.cu [hunk 2]
```diff
@@ -676,6 +676,10 @@ __device__ __host__ inline Numb operate(Numb a, Numb b) {
         case OP_AND: return new_f24(atan2f(av, bv));
         case OP_OR:  return new_f24(logf(bv) / logf(av));
         case OP_XOR: return new_f24(powf(av, bv));
+        case OP_SHL: return new_f24(sin(av + bv));
+        case FP_SHL: return new_f24(sin(av + bv));
```
Comment (developedby) [inline src/hvm.cu:680]: Same

File: src/hvm.rs [hunk 3]
```diff
@@ -407,6 +407,10 @@ impl Numb {
           OP_AND => Numb::new_f24(av.atan2(bv)),
           OP_OR  => Numb::new_f24(bv.log(av)),
           OP_XOR => Numb::new_f24(av.powf(bv)),
+          OP_SHL => Numb::new_f24((av + bv).sin()),
+          OP_SHR => Numb::new_f24((av + bv).tan()),
+          FP_SHL => Numb::new_f24((av + bv).sin()),
```
Comment (developedby) [inline src/hvm.rs:412]: and here

**Review by nicolas-abril — CHANGES_REQUESTED** [review]

File: src/ast.rs [hunk 4]
```diff
@@ -79,6 +79,8 @@ impl<'i> CoreParser<'i> {
       _ if self.try_consume("&")   => hvm::OP_AND,
       _ if self.try_consume("|")   => hvm::OP_OR,
       _ if self.try_consume("^")   => hvm::OP_XOR,
+      _ if self.try_consume("sin") => hvm::OP_SHL,
```
Comment (developedby) [inline src/ast.rs:82]: These symbols are trying to uniquely map to operators, so we shouldn't add sin and tan here. Notice how there's no pow, log and atan2.

**Review by nicolas-abril — CHANGES_REQUESTED** [review]

File: src/ast.rs (hunk 4 above)
Comment (enricozb) [reply src/ast.rs:82]: fair enough but in the context of `[` nothing else could appear there, thought it would be a bit more readable.

**Review by nicolas-abril — APPROVED** [review]

**Pull request merged**